lib/pacemaker/pcmk_rule.c

```diff
@@ -0,0 +1,291 @@
+        out->err(out, "Can't check rule %s because it does not have exactly "
+                 "one date_expression", rule_id);
+        goto done;
+    }
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    /* Then, check that it's something we actually support. */
+    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                              "@operation!='date_spec']", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max == 0) {
+        free(xpath);
+        freeXpathObject(xpath_obj);
+
+        xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                                  "@operation='date_spec' "
+                                  "and date_spec/@years "
+                                  "and not(date_spec/@moon)]", rule_id);
+        xpath_obj = xpath_search(cib_constraints, xpath);
+        max = numXpathResults(xpath_obj);
+
+        if (max == 0) {
+            rc = ENXIO;
+            out->err(out, "Rule must either not use date_spec, or use "
+                     "date_spec with years= but not moon=");
+            goto done;
+        }
+    }
+
+    match = getXpathResult(xpath_obj, 0);
+
+    /* We should have ensured this with the xpath query above, but double-
+     * checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+
+    rc = eval_date_expression(match, date, NULL);
+    out->message(out, "rule-check", rule_id, rc);
+
+done:
+    free(xpath);
+    freeXpathObject(xpath_obj);
+    return rc;
+}
+
+/*!
+ * \internal
+ * \brief Check whether each rule in a list is in effect
+ *
+ * \param[in,out] out       Output object
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date and
+ *                          time (if \c NULL, use current date and time)
+ * \param[in]     rule_ids  The IDs of the rules to check, as a <tt>NULL</tt>-
+ *                          terminated list.
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk__check_rules(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
+                  const char **rule_ids)
+{
+    pe_working_set_t *data_set = NULL;
+    int rc = pcmk_rc_ok;
+
+    CRM_ASSERT(out != NULL);
+
+    if (rule_ids == NULL) {
+        // Trivial case; every rule specified is in effect
+        return pcmk_rc_ok;
+    }
+
+    rc = rule_init(out, input, date, &data_set);
+    if (rc != pcmk_rc_ok) {
+        return rc;
+    }
+
+    for (const char **rule_id = rule_ids; *rule_id != NULL; rule_id++) {
+        int last_rc = crm_rule_check(out, data_set, *rule_id, date);
+
+        if (last_rc != pcmk_rc_ok) {
+            rc = last_rc;
+        }
+    }
+
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+// Documented in pacemaker.h
+int
+pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
+                 const char **rule_ids)
```

Comment:
The public API should be added in its own commit for the message

Reply:
Done